### PR TITLE
Adding CSI Spec v1.6.0 support for Powerflex

### PR DESCRIPTION
**Pull request by KshitijaKakde**
# Description
Adding CSI Spec v1.6.0 support for Powerflex

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| https://github.com/dell/csm/issues/905

# Checklist:

- [X] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [X] I have verified that new and existing unit tests pass locally with my changes
- [ ] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] Backward compatibility is not broken

# How Has This Been Tested?
Tested the changes by executing 2vols helm test and cert-csi

image

[root@master-1-UBC3iTe13Py54 ~]# ./cert-csi certify --cert-config /root/certify.yaml  --vsc vxflexos-snapclass
[2023-10-30 06:31:54]  INFO Starting cert-csi; ver. 0.8.1
[2023-10-30 06:31:54]  INFO Suites to run with vxflexos storage class:
[2023-10-30 06:31:54]  INFO 1. ScalingSuite {replicas: 2, volumes: 5, volumeSize: 8Gi}
[2023-10-30 06:31:54]  INFO 2. CloneVolumeSuite {pods: 2, volumes: 1, volumeSize: 8Gi}
[2023-10-30 06:31:54]  INFO 3. VolumeIoSuite {volumes: 2, volumeSize: 8Gi chains: 2-2}
[2023-10-30 06:31:54]  INFO 4. VolumeExpansionSuite {pods: 1, volumes: 1, size: 8Gi, expSize: 16Gi, block: false}
[2023-10-30 06:31:54]  INFO 5. VolumeExpansionSuite {pods: 1, volumes: 1, size: 8Gi, expSize: 16Gi, block: true}
[2023-10-30 06:31:54]  INFO 6. SnapSuite {snapshots: 3, volumeSize; 8Gi}
[2023-10-30 06:31:54]  INFO 7. ReplicationSuite {pods: 2, volumes: 5, volumeSize: 8Gi}
[2023-10-30 06:31:54]  INFO 8. MultiAttachSuite {pods: 5, rawBlock: true, size: 8Gi, accMode: ReadWriteMany}
[2023-10-30 06:31:54]  INFO Suites to run with vxflexos-xfs storage class:
[2023-10-30 06:31:54]  INFO 1. ScalingSuite {replicas: 2, volumes: 5, volumeSize: 8Gi}
[2023-10-30 06:31:54]  INFO 2. CloneVolumeSuite {pods: 2, volumes: 1, volumeSize: 8Gi}
[2023-10-30 06:31:54]  INFO 3. VolumeIoSuite {volumes: 2, volumeSize: 8Gi chains: 2-2}
[2023-10-30 06:31:54]  INFO 4. VolumeExpansionSuite {pods: 1, volumes: 1, size: 8Gi, expSize: 16Gi, block: false}
[2023-10-30 06:31:54]  INFO 5. VolumeExpansionSuite {pods: 1, volumes: 1, size: 8Gi, expSize: 16Gi, block: true}
[2023-10-30 06:31:54]  INFO 6. SnapSuite {snapshots: 3, volumeSize; 8Gi}
[2023-10-30 06:31:54]  INFO 7. ReplicationSuite {pods: 2, volumes: 5, volumeSize: 8Gi}
[2023-10-30 06:31:54]  INFO 8. MultiAttachSuite {pods: 5, rawBlock: true, size: 8Gi, accMode: ReadWriteMany}
Does it look OK? (Y)es/(n)o
-> y

 /root/.cert-csi/reports/test-run-b2aff4ca/VolumeExpansionSuite37/PodCreation.png
                        BoxPlot:
        /root/.cert-csi/reports/test-run-b2aff4ca/VolumeExpansionSuite37/PodCreation_boxplot.png
                    PodDeletion:
                        Avg: 1.615004355s
                        Min: 1.615004355s
                        Max: 1.615004355s
                        Histogram:
        /root/.cert-csi/reports/test-run-b2aff4ca/VolumeExpansionSuite37/PodDeletion.png
                        BoxPlot:
        /root/.cert-csi/reports/test-run-b2aff4ca/VolumeExpansionSuite37/PodDeletion_boxplot.png
                        EntityNumberOverTime:
        /root/.cert-csi/reports/test-run-b2aff4ca/VolumeExpansionSuite37/EntityNumberOverTime.png

[2023-10-30 06:35:06]  INFO Avg time of a run:   90.14s
[2023-10-30 06:35:06]  INFO Avg time of a del:   17.54s
[2023-10-30 06:35:06]  INFO Avg time of all:     115.08s
[2023-10-30 06:35:06]  INFO During this run 100.0% of suites succeeded

### CODE REVIEW

File: service/controller.go [hunk 1]
```diff
@@ -20,12 +20,14 @@ import (
 	"net/http"
 	"strconv"
 	"strings"
+	"sync"
```
Comment (adarsh-dell) [inline service/controller.go:23]: Can we modify Copyright years for all files that we are modifying as part of this PR?

File: service/controller.go [hunk 2]
```diff
@@ -2091,6 +2093,10 @@ func (s *service) getCapacityForAllSystems(ctx context.Context, protectionDomain
 	return capacity, nil
 }
 
+var maxVolumesSizeForArray = make(map[string]int64)
```
Comment (adarsh-dell) [inline service/controller.go:2096]: Will it be okay to add one line comment about this var?
Something like: // maxVolumesSizeForArray -  store the maxVolumesSizeForArray

File: service/controller.go (hunk 1 above)
Comment (KshitijaKakde) [reply service/controller.go:23]: Done

File: service/controller.go (hunk 2 above)
Comment (KshitijaKakde) [reply service/controller.go:2097]: @adarsh-dell review comments addressed

File: service/controller.go [hunk 3]
```diff
@@ -2132,9 +2139,75 @@ func (s *service) GetCapacity(
 			"Unable to get capacity: %s", err.Error())
 	}
 
+	systemID := ""
+	for key, value := range params {
```
Comment (adarsh-dell) [inline service/controller.go:2143]: @KshitijaKakde We have already retrieved the systemID from the request body so I believe we can remove this duplicate code here, what do you think?
image

File: service/controller.go [hunk 4]
```diff
@@ -2132,9 +2139,75 @@ func (s *service) GetCapacity(
...
+			systemID = value
+			break
+		}
+	}
+
+	maxVolSize, err := s.getMaximumVolumeSize(systemID)
+	if err != nil {
+		Log.Debug("GetMaxVolumeSize returning error ", err)
+	}
+	maxVolSizeinbps := maxVolSize * bytesInGiB
```
Comment (adarsh-dell) [inline service/controller.go:2154]: @KshitijaKakde Perform this (maxVolSizeinbps := maxVolSize * bytesInGiB) operation after if maxVolSize < 0  check.

Comment (KshitijaKakde) [reply service/controller.go]: Done

File: service/controller.go [hunk 5]
```diff
@@ -2132,11 +2139,83 @@ func (s *service) GetCapacity(
 			"Unable to get capacity: %s", err.Error())
 	}
 
+	Log.Println("defaulttttttttt", s.opts.defaultSystemID)
```
Comment (adarsh-dell) [inline service/controller.go:2142]: Typo * => Log.Println("defaulttttttttt", s.opts.defaultSystemID), Please fix this.

Comment (KshitijaKakde) [reply service/controller.go]: Done

**Review by adarsh-dell — DISMISSED** [review]
Now, Good to go from my end.

**Review by AkshaySainiDell — APPROVED** [review]
lgtm

**Review by VamsiSiddu-7 — APPROVED** [review]

**Pull request merged**